### PR TITLE
waitForXXX bugfix for nonexisting element

### LINKED ISSUE #698: waitForExist, waitForVisible - error isn't thrown when element doesn't exist
When using waitForExist or waitForVisible in 3.x, promise is incorrectly fulfilled with undefined response although element doesn't exist. 

```
var cb, options, webdriverio, client;
webdriverio = require('webdriverio');
options = {desiredCapabilities: {browserName: 'chrome'}};
client = webdriverio.remote(options);

cb = function (success) {
  return function (result) {
    console.log(success, result); // outputs 'success', undefined
    return client.end();
  };
};

client.init()
  .url('http://www.google.com/')
  .waitForExist('div.nonsense')
  .then(cb('success'), cb('fail'));
```

The code above is executed successfully and after timeout, `cb('success')` is called with no parameter - result is undefined.

### PR DESCRIPTION
**Pull request by jaroslav-kubicek**
This proposal should close #698 bug issue. Promise is now rejected if timeout expired and element still doesn't exist.

### CODE REVIEW

**Comment by christian-bromann** [comment]
I guess the problem here is that we are using catch. The waitUntil command throws if the promise inside doesn't get fulfilled in time. We are catching the error in the catch function to output a better error message. I guess this doesn't get propagated. Could you take a look?

File: lib/commands/waitForExist.js [hunk 1]
```diff
@@ -27,7 +29,13 @@ module.exports = function waitForExist(selector, ms, reverse) {
         reverse = false;
     }
 
+    var start = Date.now();
+
     return this.waitUntil(function() {
+        if (Date.now() > start + ms) {
+            // throw NoSuchElement error if no element was found before timeout
+            throw new ErrorHandler(7);
+        }
```
Comment (christian-bromann) [inline lib/commands/waitForExist.js]: waitUntil takes care of this: https://github.com/webdriverio/webdriverio/blob/master/lib/commands/waitUntil.js#L66-L72

File: lib/commands/waitForVisible.js [hunk 2]
```diff
@@ -32,7 +34,13 @@ module.exports = function waitForVisible(selector, ms, reverse) {
         reverse = false;
     }
 
+    var start = Date.now();
+
     return this.waitUntil(function() {
+        if (Date.now() > start + ms) {
+            // throw NoSuchElement error if no element was found before timeout
+            throw new ErrorHandler(7);
+        }
```
Comment (christian-bromann) [inline lib/commands/waitForVisible.js]: same here

**Comment by jaroslav-kubicek** [comment]
Ok, so I looked deep into it and it's caused by following:

In waitUntil.js, `promise.then(function(res) { ... }` is repeatedly resolved with false and because condition isn't promise, this line is called: `return defer.resolve(self.pause(250).waitUntil(condition, ms, starttime));`. 

After timeout, we are calling this:

```
    if(!timeLeft) {
        return reject(defer);
    }
```

which returns undefined...

**Comment by jaroslav-kubicek** [comment]
It's working when I change it to this:

```
if(!timeLeft) {
        reject(defer);
        return defer.promise;
 }
```

**Comment by christian-bromann** [comment]
:+1: perfect, thanks for looking into this!

**Comment by Sanjo** [comment]
@christian-bromann Can you do a patch release for this (3.1.1) please?

**Comment by adrai** [comment]
+1 but minor not patch, right?
PS. @jaroslav-kubicek well done :beer:

**Comment by cristian-smocot** [comment]
+1

This needs a minor release.
Many people use waitFor... in their scripts.

**Comment by gary-moon** [comment]
+1

This would awesome, since some of my tests were randomly succeeding when they should have been failing.

**Comment by christian-bromann** [comment]
Fixed and published with `v3.2.0`

**Comment by stoeffel** [comment]
Awesome work @jaroslav-kubicek :sparkles:

**Pull request merged**